lib/rust/ensogl/core/src/display/shape/primitive/system/cached.rs

```diff
@@ -0,0 +1,488 @@
+//!   give aliasing effects when zoomed out more than this distance.
+//! * The shapes rendered from the texture may not always gives the best quality results, see
+//!   _Limitations_ section of [`AnyCachedShape`] docs.
+//!
+//! # Usages
+//!
+//! The main idea behind cached shapes is speeding up rendering applications with many shape
+//! systems. Because each shape system is drawn with a different draw call, having e.g. a grid of
+//! icons can increase number od those, greatly worsening the application performance. But instead
+//! all the icons can be rendered to texture, and the icon grid can use only one shape system
+//! parameterized by the exact icon id.
+//!
+//! # Cached Shapes Texture
+//!
+//! As a part of the [`World`](crate::display::world::World) initialization, the exact size of the
+//! texture and shape positions is computed. The packing algorithm is defined in
+//! [`arrange_on_texture`] module. Then, the shapes are rendered to texture in
+//! [`CacheShapesPass`](crate::display::render::passes::CacheShapesPass) as soon as they are ready
+//! (their shader is compiled).
+//!
+//! The texture is rendered in
+//! [`CachdShapesTexture` display
+//! mode](crate::display::shape::primitive::glsl::codes::DisplayModes), where the RGB channels keeps
+//! color of given pixel (as one would expect), and the alpha channel contains the information about
+//! signed distance the shape boundary, linearly transformed so:
+//! * 0.5 value means distance 0.0 distance,
+//! * 0.0 is [`CACHED_TEXTURE_MAX_DISTANCE`],
+//! * 1.0 is `-CACHED_TEXTURE_MAX_DISTANCE`.
+//! As the alpha channel is capped at `0.0..=1.0` range, this will limit the stored signed distance
+//! to `-CACHED_TEXTURE_MAX_DISTANCE..=CACHED_TEXTURE_MAX_DISTANCE`, which is one of the limitations
+//! of cached shapes quality (see _Limitations_ section).
+//!
+//! # Using Cached Shapes
+//!
+//! The shapes may be read from the texture by creating [`AnyCachedShape`]. This structure can be
+//! also a parameter for shape systems defined with [`shape!`] macro. See [the docs](AnyCachedShape)
+//! for details.
+//!
+//! # Example
+//!
+//! ```rust
+//! use ensogl_core::display::shape::*;
+//!
+//!
+//! // === Defining Cached Shapes ===
+//!
+//! mod cached {
+//!     use super::*;
+//!     ensogl_core::cached_shape! { 32 x 32;
+//!                                                     // Some definition
+//! #       (_style: Style) { Circle(16.px()).into() }
+//!                                                 }
+//! }
+//!
+//! mod another_cached {
+//!     use super::*;
+//!     ensogl_core::cached_shape! { 32 x 32;
+//!                                                     // Some definition
+//! #       (_style: Style) { Circle(16.px()).into() }
+//!                                                 }
+//! }
+//!
+//!
+//! // === Using Cached Shapes ===
+//!
+//! mod shape {
+//!     use super::*;
+//!
+//!     ensogl_core::shape! {
```

Comment:
strange spacing here

Reply:
It's auto-formatter. I could not make it work or suppress in this snipped.